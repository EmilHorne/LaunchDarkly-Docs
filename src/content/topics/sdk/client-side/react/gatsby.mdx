---
path: /sdk/client-side/react/gatsby
title: Gatsby plugin
description: This topic explains where to find documentation on LaunchDarkly's Gatsby plugin, and the prerequisites for using it.
tags: ['gatsby', 'react', 'sdk', 'client-side']
published: true
---

## Overview

This topic explains where to find documentation on LaunchDarkly's [Gatsby](https://www.gatsbyjs.com/) plugin, and the prerequisites for using it. 

You can use the Gatsby plugin to integrate LaunchDarkly into your Gatsby site. This lets you use feature flags to roll out new features on your site.

## Learning about the Gatsby plugin

Full documentation for the Gatsby plugin is available on the Gatsby website.

To learn more, read [Gatsby's documentation](https://www.gatsbyjs.org/packages/gatsby-plugin-launchdarkly/) or visit the plugin's [GitHub repository](https://github.com/launchdarkly/gatsby-plugin-launchdarkly).
