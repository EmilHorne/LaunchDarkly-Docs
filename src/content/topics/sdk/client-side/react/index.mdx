---
path: /sdk/client-side/react
title: React SDK reference
description: This topic explains all of the methods available in the client-side React SDK, with detail about how they work.
published: true
---
LaunchDarkly has two distinct React SDKs. The React SDK works on web based React codebases, and the React Native SDK is built specifically to work on the Android and iOS mobile runtimes. To learn more about each SDK, read [React SDK](/sdk/client-side/react/react-web) and [React Native SDK](/sdk/client-side/react/react-native).

Additionally, we have a plugin for Gatsby based websites. The Gatsby plugin wraps the React SDK and makes it easier to use in a Gatsby site.

<ReactSdks />


