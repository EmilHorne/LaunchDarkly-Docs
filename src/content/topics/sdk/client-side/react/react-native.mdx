---
path: /sdk/client-side/react/react-native
title: React Native SDK reference
description: This topic explains all of the methods available in the React Native SDK, and explains how these methods work.
published: true
---

## Overview

This reference guide documents all of the methods available in the React Native SDK, and explains how these methods work. LaunchDarkly's SDKs are open source. The source is available in its [GitHub repository](https://github.com/launchdarkly/react-native-client-sdk) and its generated [API reference](https://launchdarkly.github.io/react-native-client-sdk/).

You can also use this SDK to clone and run a [sample application](https://github.com/launchdarkly/hello-react-native).

<Callout intent="info">
<CalloutTitle>SDK version compatibility</CalloutTitle>
<CalloutDescription>

The LaunchDarkly React Native SDK is compatible with Xcode 12.2 or higher and React Native 0.64 or higher.

If you need support for earlier versions of React Native, use older versions of the LaunchDarkly React Native SDK. To learn more, read [Releases](https://github.com/launchdarkly/react-native-client-sdk/releases).

</CalloutDescription>
</Callout>

## Getting Started

<Callout intent="warning">
<CalloutTitle>Expo usage</CalloutTitle>
<CalloutDescription>

The LaunchDarkly React Native client-side SDK is not compatible with the Expo managed workflow because the SDK uses native modules. Consider using the [bare workflow](https://docs.expo.dev/bare/exploring-bare-workflow/) instead.

</CalloutDescription>
</Callout>

After you complete the [Getting Started process](/home/getting-started), follow these instructions to start using the LaunchDarkly SDK in your React Native code.

First, install the LaunchDarkly SDK as a dependency. Add the LaunchDarkly npm dependency to your project and link it to your React Native project.

To add the LaunchDarkly npm dependency:

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'Installing with npm', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```bash
npm install --save launchdarkly-react-native-client-sdk
npx pod-install
```

</CodeTabItem>
</CodeTabs>

You must have a `Podfile` in your `ios/` directory. After you write the `Podfile`, run `pod install` in the `ios/` directory. This SDK requires Xcode 12.2 or higher and compiles against Swift 5.

After setting up the SDK, import the SDK into your project:

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
import LDClient from 'launchdarkly-react-native-client-sdk';
```

</CodeTabItem>
</CodeTabs>

After the SDK is imported, create a single, shared instance of `LDClient`, passing in the client and user configuration objects. To create a client instance, you need your environment's mobile key. This authorizes your application to connect to a particular environment within LaunchDarkly.Your mobile key is available in the Projects tab of your [Account settings page](https://app.launchdarkly.com/settings/projects). Mobile keys are not secret and you can expose them in your client-side code without risk. Never embed a server-side SDK key into a client-side application.

<Callout intent="alert">
<CalloutTitle>LDClient must be a singleton</CalloutTitle>
<CalloutDescription>

It's important to make `LDClient` a singleton. The client instance maintains internal state that allows LaunchDarkly to serve feature flags without making any remote requests. Do not instantiate a new client with every request.

</CalloutDescription>
</Callout>

The following example shows the simplest way to create a shared instance of `LDClient`:

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
let client = new LDClient();
let config = { "mobileKey": "YOUR_MOBILE_KEY" };
let user = { "key": "user_key" };

await client.configure(config, user);
```

</CodeTabItem>
</CodeTabs>

You can give `.configure` a timeout parameter. If it receives flag values before the timeout, the returned promise will resolve . Otherwise, it will be rejected.

Here is an example with a timeout of five seconds:

<CodeTabs
  defaultValue="js"
  values={[
    { label: 'JavaScript', value: 'js', },
  ]
}>
<CodeTabItem value="js">

```js
let client = new LDClient();
let config = { "mobileKey": "YOUR_MOBILE_KEY" };
let user = { "key": "user_key" };

client.configure(config, user, 5.0)
```

</CodeTabItem>
</CodeTabs>

<Callout intent="alert">
<CalloutTitle>Making feature flags available to this SDK</CalloutTitle>
<CalloutDescription>

You must make feature flags available to mobile SDKs before the SDK can evaluate those flags. If an SDK tries to evaluate a feature flag that is not available, the user will receive the default value for that flag.

To make a flag available to this SDK, check the **SDKs using Mobile key** checkbox during flag creation, or on the flag's Settings tab. To make all of a project's flags available to this SDK by default, check the **SDKs using Mobile key** checkbox in your project [Settings](https://app.launchdarkly.com/settings/projects).

</CalloutDescription>
</Callout>

## Background fetch

The React Native SDK defaults to the background fetch behavior for the platform it's running on.

## Shutting down

Shut down the client when your application terminates. To learn more, read [Shutting down](/sdk/features/shutdown#react-native).

## Data collection

To learn more about data collection within this SDK and implications on submissions to the Apple App Store, read [Apple App Store data collection policy](/sdk/concepts/apple-app-store).

## Supported features

This SDK supports the following features:

* [Aliasing users](/sdk/features/aliasing-users#react-native)
* [Configuration](/sdk/features/config#react-native)
* [Evaluating flags](/sdk/features/evaluating#react-native)
* [Flag evaluation reasons](/sdk/features/evaluation-reasons#react-native)
* [Flushing events](/sdk/features/flush#react-native)
* [Getting all flags](/sdk/features/all-flags#react-native)
* [Identifying and changing users](/sdk/features/identify#react-native)
* [Monitoring SDK status](/sdk/features/monitoring#react-native)
* [Offline mode](/sdk/features/offline-mode#react-native)
* [Relay Proxy configuration, using proxy mode](/sdk/features/relay-proxy-configuration/proxy-mode#react-native)
* [Sending custom events](/sdk/features/events#react-native)
* [Shutting down](/sdk/features/shutdown#react-native)
* [Subscribing to flag changes](/sdk/features/flag-changes#react-native)
* [User configuration](/sdk/features/user-config#react-native)
