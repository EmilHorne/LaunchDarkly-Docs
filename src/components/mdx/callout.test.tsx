import React from 'react'
import { render } from '@testing-library/react'

import Callout, { CalloutDescription, CalloutTitle } from './callout'

describe('callout', () => {
  test('renders', () => {
    const { asFragment } = render(
      <Callout intent="primary">
        <CalloutTitle>Adding users to LaunchDarkly</CalloutTitle>
        <CalloutDescription>
          You do not have to send users to LaunchDarkly in advance. <a href="example.com">Check out this page</a>. You
          can target them with feature flags have LaunchDarkly accounts of their own. Users appear in the dashboard
          automatically after they encounter feature flags.
        </CalloutDescription>
      </Callout>,
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
