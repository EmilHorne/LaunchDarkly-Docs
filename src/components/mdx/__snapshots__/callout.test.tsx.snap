// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`callout renders 1`] = `
<DocumentFragment>
  .emotion-0 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
}

.emotion-1 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-3 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  margin-bottom: 16px;
  margin-right: 128px;
  font-size: 24px;
  line-height: regular;
}

.emotion-4 {
  box-sizing: border-box;
  margin: 0;
  min-width: 0;
  font-size: 20px;
  line-height: spaced;
  margin-bottom: 4px;
  overflow-wrap: break-word;
}

.emotion-4 p {
  font-size: 20px;
  line-height: spaced;
  margin-bottom: 4px;
  overflow-wrap: break-word;
}

<div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-0"
      >
        <span
          class="emotion-3"
        >
          Adding users to LaunchDarkly
        </span>
        <span
          class="emotion-4"
        >
          You do not have to send users to LaunchDarkly in advance. 
          <a
            href="example.com"
          >
            Check out this page
          </a>
          . You can target them with feature flags have LaunchDarkly accounts of their own. Users appear in the dashboard automatically after they encounter feature flags.
        </span>
      </div>
      <div>
        Icon: checkbox-marked-circle
      </div>
    </div>
  </div>
</DocumentFragment>
`;
